runtime/vm/ExtendedMessageNPE.cpp

```diff
@@ -0,0 +1,201 @@
+			msgLen = j9str_printf(PORTLIB, NULL, 0, VM_NPE_MONITORENTER);
+			npeMsg = (char *)j9mem_allocate_memory(msgLen+ 1, OMRMEM_CATEGORY_VM);
+			j9str_printf(PORTLIB, npeMsg, msgLen, VM_NPE_MONITORENTER);
+#undef VM_NPE_MONITORENTER
+			break;
+		case JBmonitorexit:
+#define VM_NPE_MONITOREXIT "Cannot exit synchronized block"
+			msgLen = j9str_printf(PORTLIB, NULL, 0, VM_NPE_MONITOREXIT);
+			npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
+			j9str_printf(PORTLIB, npeMsg, msgLen, VM_NPE_MONITOREXIT);
+#undef VM_NPE_MONITOREXIT
+			break;
+		case JBgetfield: /* FALLTHROUGH */
+		case JBputfield:
+			{
+				U_16 index = PARAM_16(bcCurrentPtr, 1);
+				UDATA cpType = J9_CP_TYPE(J9ROMCLASS_CPSHAPEDESCRIPTION(romClass), index);
+
+				Trc_VM_GetCompleteNPEMessage_Field_Index(vmThread, index);
+				if (J9CPTYPE_FIELD == cpType) {
+					J9ROMConstantPoolItem *constantPool = J9_ROM_CP_FROM_ROM_CLASS(romClass);
+					J9ROMConstantPoolItem *cpItem = constantPool + index;
+					J9ROMNameAndSignature *nameAndSig = J9ROMFIELDREF_NAMEANDSIGNATURE((J9ROMFieldRef *) cpItem);
+					J9UTF8 *name = J9ROMNAMEANDSIGNATURE_NAME(nameAndSig);
+
+					if (JBputfield == bcCurrent) {
+#define VM_NPE_PUTFIELD "Cannot assign field \"%2$.*1$s\""
+						nlsMessage = VM_NPE_PUTFIELD;
+#undef VM_NPE_PUTFIELD
+					} else {
+#define VM_NPE_GETFIELD "Cannot read field \"%2$.*1$s\""
+						nlsMessage = VM_NPE_GETFIELD;
+#undef VM_NPE_GETFIELD
+					}
+					msgLen = j9str_printf(PORTLIB, NULL, 0, nlsMessage, J9UTF8_LENGTH(name), J9UTF8_DATA(name));
+					npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
+					j9str_printf(PORTLIB, npeMsg, msgLen, nlsMessage, J9UTF8_LENGTH(name), J9UTF8_DATA(name));
+				} else {
+					Trc_VM_GetCompleteNPEMessage_UnexpectedCPType(vmThread, cpType, bcCurrent);
+				}
+			}
+			break;
+		case JBinvokehandle: /* FALLTHROUGH */
+		case JBinvokehandlegeneric: /* FALLTHROUGH */
+		case JBinvokeinterface2: /* FALLTHROUGH */
+		case JBinvokeinterface: /* FALLTHROUGH */
+		case JBinvokespecial: /* FALLTHROUGH */
+		case JBinvokedynamic: /* FALLTHROUGH */
+		case JBinvokevirtual: /* FALLTHROUGH */
+		case JBinvokestatic:
+			{
+				U_16 index = 0;
+				J9ROMNameAndSignature *nameAndSig = NULL;
+				J9UTF8 *sig = NULL;
+				J9UTF8 *name = NULL;
+				J9UTF8 *definingUTF = NULL;
+				J9ROMMethodRef *romMethodRef = NULL;
+				J9ROMConstantPoolItem *constantPool = J9_ROM_CP_FROM_ROM_CLASS(romClass);
+
+				if (JBinvokeinterface2 == bcCurrent) {
+					index = PARAM_16(bcCurrentPtr + 2, 1); /* get JBinvokeinterface instead */
+				} else {
+					index = PARAM_16(bcCurrentPtr, 1);
+				}
+				Trc_VM_GetCompleteNPEMessage_Invoke_Index(vmThread, index);
+				romMethodRef = (J9ROMMethodRef *)&constantPool[index];
+				nameAndSig = J9ROMMETHODREF_NAMEANDSIGNATURE(romMethodRef);
+				sig = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSig);
+				name = J9ROMNAMEANDSIGNATURE_NAME(nameAndSig);
```

Comment:
The locals can be declared at first use as this is a cpp file.  Saves initializing them to null and shortens the life time they need to be reasoned about

Reply:
Yeah, should use this advantage, updated.